### ISSUE TITLE
Next version

### ISSUE THREAD

**Issue by pnevyk** [datamole-ai/gomez#9]
Ever since I discovered the [faer](https://github.com/sarah-ek/faer-rs) linear algebra library I wanted to rewrite gomez into it. For two reasons:

1. Performance. The [benchmarks](https://faer-rs.github.io/bench-f64/) for faer are very impressive. Moreover, the low level API allows to do operations, for which I am now forced to use cloning, in place.
2. API. nalgebra is very flexible with the matrix storage, but it also means that it's very trait-heavy (our public API is unnecessarily [complex](https://github.com/datamole-ai/gomez/blob/ee6411ad2c648673a5b3054f5fda8832ddb716aa/src/core/system.rs#L64-L71) due to that, and the implementations of solvers are even [worse](https://github.com/datamole-ai/gomez/blob/ee6411ad2c648673a5b3054f5fda8832ddb716aa/src/solver/trust_region.rs#L224-L228)). Interface when using faer will be much simpler.

Half a year ago I actually spent some time on implementing the algorithms using faer, and after initial struggles I managed to do most of them (and some more, and in a much more modular structure). The code that I have is for an old version of faer and a lot of changes were made since that, but they should be mostly cosmetic.

The plan is as follows:

1. Make some improvements to the API (more on that below).
2. Release 0.5.0.
3. Migrate everything to faer.
4. Release 0.6.0.
5. For a limited time period, make patches to 0.5.x branch if needed. This would be for people who want to continue using nalgebra.

Before fully migrating to faer, I would like to do some unrelated (or partly related) improvements to the API. Now I can think of these:

- [ ] Make function optimization a first-class citizen. Rename the `next` methods in `Solver` and `Optimizer` traits to `solve_next` and `opt_next`, respectively, so that they don't clash (see dd6524a).
- [ ] Remove `apply_eval` from the `Function` trait. A problem is either a system or function, not both. I guess this was a work around for being able to "solve" some systems by algorithms that only implemented the `Optimizer` trait. Instead of `apply_eval`, all `Optimizer`s should also implement the `Solver` trait (not automatically, but as a guideline).
- [ ] Remove `Dim` associated type from the `Problem` trait. faer supports only dynamically sized matrices. Removing the type now should be an incremental change towards to final migration. Move the `dim` method from the `Problem` trait to the `Domain` type and make its return type `usize`.
- [ ] Add `jacobian` method to the `System` trait and `gradient` and `hessian` methods to the `Function` trait and provide default implementations using finite difference approximations.
- [ ] Remove `ProblemError` and don't return `Result` in `System::eval` and `Function::apply`. Encountering invalid dimensionality is a bug in the solver/optimizer not respecting the value in `Domain::dim` and so `panic!` is appropriate. Invalid values (infinity, NaN) should be detected by the solver. This will simplify APIs a bit and might also result in faster & smaller code (less branches). Transitively, other error types and corresponding `Result` returning methods will also be removed (`JacobianError`, etc.).
- [ ] Rename `Problem::Scalar` associated type to `Problem::Field` (seems like a better name considering its usage in Rust ecosystem). Introduce custom `RealField` trait having nalgebra's (later faer's) `RealField` as a super trait, and make `Problem::Field` type be constrained by our `RealField` trait. Move `EPSILON_CBRT` and `EPSILON_SQRT` constants to this trait (and use proper values for `f32`).
- [ ] Remove `Variable` and `VariableBuilder` and add specific constructors to `Domain` for unconstrained and rectangular domain. This will make `Domain` more flexible (we might add support for other domain types, for example spherical domain). But user should still be able to somehow specify custom scale/magnitude for variables.
- [ ] Make `Problem::domain` return `&Domain` instead of `Domain`. This will shift the responsibility of keeping the domain instance from the call-site (before creating the solver and throughout the solving process) to the problem type itself. Thanks to that, many functions will not need separate parameter for the domain.
- [ ] Remove cuckoo search algorithm. I don't think it provides much value, it was a relic from my experimentation in the past. I believe that LIPO or a more widely-used meta-heuristics algorithm (e.g., differential evolution) should be our focus.
- [ ] Remove the `population` module. As cuckoo search will be removed and it is currently the only population-based algorithm in gomez, there is no point in having this module in gomez (now). We may reintroduce it later, potentially in a different shape.
- [ ] Remove `FunctionResultExt` and `VectorDomainExt`. These are poor API.
- [ ] Hide the `testing` module behind a feature flag.
- [ ] Add `TestFunction` counterpart to the `TestSystem` in the `testing` module.
- [ ] Rename `SolveError` in `testing` module to something more generic (e.g., `TestError`) so the naming makes more sense in the context of optimization.
- [ ] Make `Solver` and `Optimizer` the only things exported from the `prelude`. Since `Problem` and `System`/`Function` are traits to be implemented by the user, it is probably better to explicitly import them.
- [ ] Introduce `SolveIter<'a>` and `OptimizeIter<'a>` which implement `Iterator<Item = Result<('a x, 'a fx), S::Error>>` (no lifetime `'a` for fx in optimization case). Add `solve(solver, problem, x, fx)` and `optimize(optimizer, problem, x)` methods to the `Solver` and `Optimizer` traits, respectively, with default implementation that will simply construct the corresponding iterator and return it. This will be the high-level API, which still allows quite flexible control over the process (e.g., `iter.take(max_iters).find(|result| result.map(|(x, fx)| fx.norm() <= 1e-6).unwrap_or(true))`).

After all these changes are done and 0.5.0 is released, then I will migrate the whole library from nalgebra to faer.

- [ ] Migrate to faer.

**Comment by jolars** [comment]
I ran some benchmarks on this in my basin library, and I dont think you should expect any great improvement (if any at all).

https://basin.rs/benchmarks/backends/

**Comment by pnevyk** [comment]
Thank you for including gomez in the benchmarks. I checked your library and it looks super cool! It must have taken really significant effort.

I don't have a lot of time to check the source code in detail. My expectation for higher performance comes from some measurements in the past, where the main algorithm of gomez (called trust region in the library) spends a lot of time in QR decomposition and allocation of a full matrix on every step due to limitations of nalgebra API. Given the [faer's own benchmarks](https://faer.veganb.tw/benchmarks/peddie6146/), the performance of QR decomposition should be considerably faster than of nalgebra and the low level API would allow me to reuse matrix allocations. This should play a role especially for problems with many dimensions (at least hundreds).

That said, I haven't run any direct performance comparison between the two so my assumptions may turn out false.

By the way, I know that the source code is quite a mess (that's one of the main reasons I wanted to refactor it), but feel free to take inspiration from the [implementation of the trust region algorithm](https://github.com/datamole-ai/gomez/blob/54310ffdb29493b42b862ec66f89b738f1603f43/src/algo/trust_region.rs) in gomez. I researched a lot of sources for numerical tricks which make a difference in practice compared to textbook definitions of the algorithms.

**Comment by jolars** [comment]
Yes, I'm not sure why there's a discrepancy. Even at the lower dimensions (where the faer and basin benchmarks overlap) it looks like faer should be beating the competition most of the time, but nalgebra is usually doing better, even in the CMA-ES (eigendecompositions) and LM (QR factoriziations) solvers, so I'm not sure what's going on.

Thanks for the pointer to your TrustRegion work! That's on my todo list, so it's good timing! :) I'll be sure to reference your work if I end up porting any of it.
